libs/shared/ui-components/src/SocialCard/SocialCard.tsx

```diff
@@ -0,0 +1,96 @@
+import { useRouter } from 'next/router';
+
+import { DomainVariant } from '@quansight/shared/types';
+
+import { TSocialCardProps } from './types';
+
+export const SocialCard: FC<TSocialCardProps> = ({
+  title,
+  description,
+  variant,
+  twitterImage,
+  ogImage,
+  alt,
+  summaryLargeImage = true,
+}) => {
+  const { asPath } = useRouter();
+
+  /*
+    url is constructed starting from https:// because 'NEXT_PUBLIC_VERCEL_URL' doesn't provide the protocol(https)
+    and for og:url requires absolute URL which includes the protocol and domain name
+    using the prefix NEXT_PUBLIC is used to indicate that the environment variable is public,
+    meaning that it can be accessed by the browser.
+    'NEXT_PUBLIC_VERCEL_URL' is set by Vercel when the application is deployed, and it can be accessed by the application code.
+    This kind of URL is need because of og:url needs absolute path and allows it to use this as the correct canonical tag
+    Also twitter doesn't have something similar to og:url like twitter:url.
+   */
+  const url = `https://${process.env['NEXT_PUBLIC_VERCEL_URL']}${asPath}`;
+
+  const defaultTwitterImage = (variant: DomainVariant): string => {
+    switch (variant) {
+      case DomainVariant.Quansight:
+        return 'https://a.storyblok.com/f/147759/1200x630/8b20f86b2a/quansight-logo-opengraph-size.png';
+      case DomainVariant.Labs:
+        return 'https://a.storyblok.com/f/152463/900x472/b289c484e3/qslabs-logo-opengraph-size.png';
+      default:
+        return '';
+    }
+  };
+  const defaultOgImage = (variant: DomainVariant): string => {
+    switch (variant) {
+      case DomainVariant.Quansight:
+        return 'https://a.storyblok.com/f/147759/1200x630/8b20f86b2a/quansight-logo-opengraph-size.png';
+      case DomainVariant.Labs:
+        return 'https://a.storyblok.com/f/152463/900x472/b289c484e3/qslabs-logo-opengraph-size.png';
+      default:
+        return '';
+    }
+  };
+  const defaultAlt = (variant: DomainVariant): string => {
+    switch (variant) {
+      case DomainVariant.Quansight:
+        return 'Quansight logo';
+      case DomainVariant.Labs:
+        return 'Quansight Labs logo';
+      default:
+        return '';
+    }
+  };
+  return (
+    <Head>
+      {/* twitter */}
+      <meta
+        name="twitter:card"
+        content={summaryLargeImage ? 'summary_large_image' : 'summary'}
+      />
+      <meta name="twitter:site" content="@quansightai" />
+      <meta name="twitter:title" content={title} />
+      <meta name="twitter:description" content={description} />
+      <meta
+        name="twitter:image"
+        content={twitterImage || defaultTwitterImage(variant)}
+      />
+      {/*
+        If `twitterImage` is specified but `alt` is not, we use blank alt-text because
+        it's better to have no alt-text than incorrect alt-text.
+       */}
+      <meta
+        name="twitter:alt"
+        content={alt || twitterImage ? '' : defaultAlt(variant)}
+      />
+
```

Comment:
I think is missing a closing curly brace

Reply:
@bskinn @gabalafou 

For the suggested code [above](https://github.com/Quansight/Quansight-website/pull/733#discussion_r1214743626)
consider the case when we have an `alt` but no `twitterImage` then what will be shown is a Quansight logo or Labs logo with the wrong `alt`. I believe if we don't have twitterImage we should always show the defaultAlt because we're showing  Quansight logo or Labs logo if we don't have one any way.

to summarize:

| | `twitterImage` | no `twitterImage` |
| -- | ---- | ---------- |
| `alt` | show `alt` | `defaultAlt` |
| no `alt`| don't have `alt` tag | `defaultAlt`|


```suggestion
     {(alt || !twitterImage) && (
        <meta
          name="twitter:alt"
          content={twitterImage ? alt : defaultAlt(variant)}
        />
      )}

```

Reply:
Yeah, that looks right to me.

However, I'm starting to think that the fact that it's taken us this many cycles to nail down this logic suggests that there's something wrong or unintuitive about the API for this React component.

Maybe we got a little bit ahead of ourselves. As far as I can tell, we're not even using these other properties anywhere at the moment. Maybe we should have started with the smallest thing first, just the title, description, and variant props, and added the other props later when we have code that uses them. But since we're already here...

My understanding is that with this component we're basically trying to support two use cases.

1. Social card with generic, default image (Quansight or Quansight Labs logo)
2. Social card with a content-specific image

The first case is easy; it's the one we see in this PR:

```tsx
<SocialCard title="..." description="..." variant={DomainVariant[...]} />
```

The second case is a bit more complicated. But I'm guessing that 90%+ of the time it should look like this:

```tsx
<SocialCard
  title="..."
  description="..."
  variant={...}
  twitterImage="https://..."
  {/* ogImage same as twitterImage, just different size */}
  ogImage="https://..." 
  alt="same alt text for both images because they're both the same, just different sizes"
/>
```

Everything else is an edge case right?

So maybe the simplest implementation for now is to not support any of the edge cases. Which we can accomplish by changing the type:

```ts
type TSocialCardPropsBase = {
  title: string;
  description: string;
  variant: DomainVariant;
};

type TSocialCardPropsDefaultImage = TSocialCardPropsBase;

// If you supply a customized social card image, you must supply the same image
// for both Twitter and Open Graph, and it must have alt text.
type TSocialCardPropsCustomizedImage = TSocialCardPropsBase & {
  twitterImage: string;
  ogImage: string;
  alt: string;
};

export type TSocialCardProps = 
  TSocialCardPropsDefaultImage |
  TSocialCardPropsCustomizedImage;
```

Then you can write the body of the React component code like so:

```tsx
      <meta
        name="twitter:image"
        content={twitterImage || defaultTwitterImage(variant)}
      />
      <meta
        name="twitter:alt"
        content={alt || defaultAlt(variant)}
      />
      <meta property="og:image" content={ogImage || defaultOgImage(variant)} />
      <meta
        property="og:image:alt"
        content={alt || defaultAlt(variant)}
      />
```

You can use simplified logic because the type-checker assures that you're either in the first case using default images and alt text, or you're in the second case where you're using custom images (both Twitter and Open Graph) and alt text.

I don't really know if we need to support the case where somebody provides a Twitter image but not alt text (at Quansight we want to be disciplined about providing alt text). Or the case where somebody provides alt text but no Twitter image—except perhaps in the case where somebody has an image and alt text for Open Graph but not Twitter?  But do we really need to support that? If we want to support the use case where somebody provides a custom image (plus alt text) for Open Graph but not Twitter, or vice versa, then that should probably be reflected in our React component's API. In other words, it should define both `twitterAlt` and `ogImageAlt`.

Thoughts?

Reply:
I like your proposed refactoring of the types, @gabalafou. I agree, that either all of these should be present together, or absent together. I originally was thinking about this differently, but I'm aligned with you now.

One thing to add, though: `TSocialCardPropsCustomizedImage` will need an additional e.g. `usingTwitterLargeImage: boolean;` property, to allow the user of the component to indicate which size of twitter card image that `twitterImage` meets.

(The statement `{/* ogImage same as twitterImage, just different size */}` is not correct when using the small-image `summary` Twitter card type.)

---

> we're not even using these other properties anywhere at the moment

No, but the next step I'd planned after this PR merges was to implement use of them elsewhere: to allow custom social card props on blog posts and general site pages.

---

In terms of future code changes:

- Blog posts
  - Change to types for the two sites:
    - Labs: Add a new `socialcard?:` to [`TPostMeta`][labs tpostmeta] in `/apps/labs/types/storyblok/bloks/posts.ts`
    - Consulting: Add a new `socialcard?:` to [`TPostMeta`][cons tpostmeta] in `/apps/consulting/types/storyblok/bloks/blogPost.ts`
    - More inconsistency between the two sites, in this filename 🫤
  - Because of these types that we can enforce, a failure to provide all four of the needed metadata fields in a post header should lead to a broken build.
- Regular (Storyblok) site pages
  - The SB schema for `Page` will need to be updated with these four fields.
  - I _think_ any validation to require that all four fields be present or absent together will have to happen on the Storyblok side.
    - The page metadata is [pulled from the page's `data.content`][data.content], which is pulled from the Apollo query.
    - Given that, e.g., `data.content.description` is optional in Storyblok and there's no indication I can see of its being optional in [the Apollo query], I don't think we can enforce this on the codebase side.
    - So... one option might be to implement a `SocialCard` component in Storyblok? Where its presence on a `Page` is optional, but all four fields within it are required?


[cons tpostmeta]: https://github.com/Quansight/Quansight-website/blob/674832b40fac9f17f54daa566642ab3c238e7ad6/apps/consulting/types/storyblok/bloks/blogPost.ts#L19-L29
[labs tpostmeta]: https://github.com/Quansight/Quansight-website/blob/05a6d3c818017eca3f2186233275ce27e98580d9/apps/labs/types/storyblok/bloks/posts.ts#L28-L38

[data.content]: https://github.com/Quansight/Quansight-website/blob/674832b40fac9f17f54daa566642ab3c238e7ad6/apps/consulting/pages/index.tsx#L33-L37

[the apollo query]: https://github.com/Quansight/Quansight-website/blob/674832b40fac9f17f54daa566642ab3c238e7ad6/libs/shared/storyblok-sdk/src/api/queries/common/page.graphql#L12-L20